xdp/src/netlink.rs

```diff
@@ -0,0 +1,601 @@
+    pub family: u8,
+    pub dst_len: u8,
+}
+
+#[repr(C)]
+#[allow(non_camel_case_types)]
+struct rtmsg {
+    rtm_family: u8,
+    rtm_dst_len: u8,
+    rtm_src_len: u8,
+    rtm_tos: u8,
+    rtm_table: u8,
+    rtm_protocol: u8,
+    rtm_scope: u8,
+    rtm_type: u8,
+    rtm_flags: u32,
+}
+
+#[repr(C)]
+struct RouteRequest {
+    header: nlmsghdr,
+    rtm: rtmsg,
+}
+
+fn parse_ip_address(data: &[u8], family: u8) -> Option<IpAddr> {
+    match family as i32 {
+        AF_INET if data.len() == 4 => Some(IpAddr::V4(Ipv4Addr::new(
+            data[0], data[1], data[2], data[3],
+        ))),
+        AF_INET6 if data.len() == 16 => {
+            let mut segments = [0u16; 8];
+            for i in 0..8 {
+                segments[i] = ((data[i * 2] as u16) << 8) | (data[i * 2 + 1] as u16);
+            }
+            Some(IpAddr::V6(Ipv6Addr::from(segments)))
+        }
+        _ => None,
+    }
+}
+
+pub fn netlink_get_routes(family: u8) -> Result<Vec<RouteEntry>, io::Error> {
+    let sock = NetlinkSocket::open()?;
+
+    // Safety: RouteRequest is POD
+    let mut req = unsafe { mem::zeroed::<RouteRequest>() };
+
+    let nlmsg_len = mem::size_of::<nlmsghdr>() + mem::size_of::<rtmsg>();
+    req.header = nlmsghdr {
+        nlmsg_len: nlmsg_len as u32,
+        nlmsg_flags: (NLM_F_REQUEST | NLM_F_DUMP) as u16,
+        nlmsg_type: RTM_GETROUTE,
+        nlmsg_pid: 0,
+        nlmsg_seq: 1,
+    };
+
+    req.rtm.rtm_family = family;
+    req.rtm.rtm_table = RT_TABLE_MAIN;
+
+    sock.send(&bytes_of(&req)[..req.header.nlmsg_len as usize])?;
+
+    let mut routes = Vec::new();
+
+    for msg in sock.recv()? {
+        if msg.header.nlmsg_type != RTM_NEWROUTE {
+            continue;
+        }
+
+        if msg.data.len() < mem::size_of::<rtmsg>() {
+            continue;
+        }
+
+        let Some(route) = parse_rtm_newroute(msg) else {
+            continue;
+        };
+
+        routes.push(route);
+    }
+
+    Ok(routes)
+}
+
+pub fn parse_rtm_newroute(msg: NetlinkMessage) -> Option<RouteEntry> {
+    let rt_msg = unsafe { ptr::read_unaligned(msg.data.as_ptr() as *const rtmsg) };
+    let Ok(attrs) = parse_attrs(&msg.data[mem::size_of::<rtmsg>()..]) else {
+        return None;
+    };
+    let mut route = RouteEntry {
+        destination: None,
+        gateway: None,
+        pref_src: None,
+        out_if_index: None,
+        in_if_index: None,
+        priority: None,
+        table: None,
+        protocol: rt_msg.rtm_protocol,
+        scope: rt_msg.rtm_scope,
+        type_: rt_msg.rtm_type,
+        family: rt_msg.rtm_family,
+        dst_len: rt_msg.rtm_dst_len,
+    };
+    if let Some(dst_attr) = attrs.get(&RTA_DST) {
+        route.destination = parse_ip_address(dst_attr.data, rt_msg.rtm_family);
+    }
+    if let Some(gateway_attr) = attrs.get(&RTA_GATEWAY) {
+        route.gateway = parse_ip_address(gateway_attr.data, rt_msg.rtm_family);
+    }
+    if let Some(oif_attr) = attrs.get(&RTA_OIF) {
```

Comment:
maybe a helper function would be nice here?

Reply:
not sure what you mean? helper function that does what?